/**
 * Copyright (c) 2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
package proto_panda;

import "assemblyIns.proto";
import "assemblyLabel.proto";
import "assemblyType.proto";
import "assemblyDebug.proto";
import "assemblyFileLocation.proto";
import "ideHelpers.proto";
import "meta.proto";

message CatchBlock {
    bytes whole_line = 1;
    bytes exception_record = 2;
    bytes try_begin_label = 3;
    bytes try_end_label = 4;
    bytes catch_begin_label = 5;
    bytes catch_end_label = 6;
}

message TryCatchInfo {
    message TryCatchLabel {
        bytes key = 1;
        uint64 value = 2;
    }
    message TryCatchMap {
        bytes key = 1;
        repeated CatchBlock catch_blocks = 2;
    }
    repeated TryCatchLabel try_catch_labels = 1;
    repeated TryCatchMap try_catch_map = 2;
    repeated bytes try_catch_order = 3;
}

message Parameter {
    Type type = 1;
    ParamMetadata metadata = 2;
}

message Function {
    message LabelTable {
        bytes key = 1;
        Label value = 2;
    }

    bytes name = 1;
    uint32 language = 2;
    FunctionMetadata metadata = 3;
    repeated LabelTable label_table = 4;
    repeated Ins ins = 5;
    repeated LocalVariable local_variable_debug = 6;
    bytes source_file = 7;
    bytes source_code = 8;
    repeated CatchBlock catch_blocks = 9;
    int64 value_of_first_param = 10;
    uint64 regs_num = 11;
    repeated Parameter params = 12;
    bool body_presence = 13;
    Type return_type = 14;
    SourceLocation body_location = 15;
    optional FileLocation file_location = 16;
}

